src/hotspot/share/opto/chaitin.cpp

```diff
@@ -1510,29 +1522,38 @@ OptoReg::Name PhaseChaitin::bias_color( LRG &lrg, int chunk ) {
     lrg.Remove(reg);
     OptoReg::Name reg2 = lrg.mask().find_first_elem();
     lrg.Insert(reg);
-    if( OptoReg::is_reg(reg2))
+    // The below check is a direct translation of an older version of this code
+    // where the chunk/offset was represented directly (rather than internally
+    // in the register masks). I suspect the check could be replaced by
+    // OptoReg::is_reg(reg2), meaning that we only want to alternate when
+    // dealing with registers in the first "chunk" (and not stack slots). But,
+    // this does change behavior compared to the old version (hence my
+    // hesitation).
+    if (OptoReg::is_valid(reg2) &&
+        OptoReg::is_reg(reg2 - lrg.mask().offset_bits())) {
```

Comment:
I agree that this was probably an oversight in the original code. For simplicity I suggest to replace the check with just `OptoReg::is_reg(reg2)` as you suggest, explicitly limiting the scope of the alternation heuristic to physical registers. I compared the overall effectiveness of post-allocation copy removal (as summarized by `-XX:+PrintOptoStatistics`) between this changeset and your proposed simplification and I cannot see any significant difference. I really wonder if the entire alternation heuristic really has any positive measurable effect, but that investigation belongs to another RFE.

Reply:
Thanks for comparing! Now changed.